### PR TITLE
Open/close functionality

### PR DESCRIPTION
**Pull request by danaack**
Open and close functionality has been implemented for open address file hash, open address hash, and skip list.

### CODE REVIEW

File: src/dictionary/dictionary.c [hunk 1]
```diff
@@ -233,18 +234,62 @@ dictionary_compare_null_terminated_string(
...
+			config->id, 0, config->type, config->key_size, config->value_size, config->dictionary_size
+		};
+
+		if (err_ok != dictionary_open(&bpp_tree_handler, &bpp_dict, &bpp_config)) {
+			return err_dictionary_initialization_failed;
+		}
+
+		dictionary_build_predicate(&predicate, predicate_all_records);
+		dictionary_find(&bpp_dict, &predicate, &cursor);
+		record.key		= (ion_key_t) malloc((size_t) config->key_size);
```
Comment (graemedouglas) [inline src/dictionary/dictionary.c]: Don't cast those malloc's!

File: src/dictionary/open_address_file_hash/open_address_file_hash.c [hunk 2]
```diff
@@ -34,8 +53,26 @@ oafh_initialize(
...
+	char addr_filename[20];
+
 	/* open the file */
-	hashmap->file						= fopen(TEST_FILE, "w+b");
+	oafdict_get_addr_filename(id, addr_filename);
+
+	if (ion_fexists(addr_filename)) {
+		hashmap->file	= fopen(addr_filename, "r+b");
+
+#if defined(ARDUINO)
```
Comment (graemedouglas) [inline src/dictionary/open_address_file_hash/open_address_file_hash.c]: These three lines need to be deleted.

File: src/dictionary/open_address_file_hash/open_address_file_hash.c [hunk 3]
```diff
@@ -34,8 +53,26 @@ oafh_initialize(
...
+	hashmap->compute_hash				= (*hashing_function);	/* Allows for binding of different hash functions
+																depending on requirements */
+
+	char addr_filename[20];
+
 	/* open the file */
-	hashmap->file						= fopen(TEST_FILE, "w+b");
+	oafdict_get_addr_filename(id, addr_filename);
+
+	if (ion_fexists(addr_filename)) {
```
Comment (graemedouglas) [inline src/dictionary/open_address_file_hash/open_address_file_hash.c]: We should simply open this and see if the file is equal to NULL.

File: src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c [hunk 4]
```diff
@@ -8,6 +8,24 @@
...
+
+void
+oafdict_get_addr_filename(
+	ion_dictionary_id_t id,
+	char				*str
+) {
+	sprintf(str, "%d.oaf", id);
+}
+
+void
```
Comment (graemedouglas) [inline src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c]: Delete this function if its not used.

File: src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c [hunk 5]
```diff
@@ -399,3 +419,43 @@ oafdict_scan(
...
+	ion_dictionary_handler_t		*handler,
+	ion_dictionary_t				*dictionary,
+	ion_dictionary_config_info_t	*config,
+	ion_dictionary_compare_t		compare
+) {
+	ion_err_t error;
+
+	error = oafdict_create_dictionary(config->id, config->type, config->key_size, config->value_size, config->dictionary_size, compare, handler, dictionary);
+
+	if (err_ok != error) {
```
Comment (graemedouglas) [inline src/dictionary/open_address_file_hash/open_address_file_hash_dictionary_handler.c]: This could be simplified to just

```
return error;
```

**Comment by Stickerpants** [comment]
LGTM

**Comment by wpenson** [comment]
LGTM

**Pull request merged**